macq/extract/locm.py

```diff
@@ -0,0 +1,802 @@
+""".. 'include':: ../../docs/templates/extract/locm.md"""
+
+from collections import defaultdict
+from dataclasses import asdict, dataclass
+from pprint import pprint
+from typing import Dict, List, NamedTuple, Optional, Set, Tuple, Union
+from warnings import warn
+
+from macq.trace.action import Action
+from macq.trace.fluent import PlanningObject
+
+from ..observation import ActionObservation, Observation, ObservedTraceList
+from . import LearnedLiftedAction, Model
+from .exceptions import IncompatibleObservationToken
+from .learned_fluent import LearnedLiftedFluent
+from .model import Model
+
+
+@dataclass
+class AP:
+    """Action.Position (of object parameter). Position is 1-indexed."""
+
+    action: Action
+    pos: int  # NOTE: 1-indexed
+    sort: int
+
+    def __repr__(self) -> str:
+        return f"{self.action.name}.{self.pos}"
+
+    def __hash__(self):
+        return hash(self.action.name + str(self.pos))
+
+    def __eq__(self, other):
+        return hash(self) == hash(other)
+
+
+class StatePointers(NamedTuple):
+    start: int
+    end: int
+
+    def __repr__(self) -> str:
+        return f"({self.start} -> {self.end})"
+
+
+Sorts = Dict[str, int]  # {obj_name: sort}
+APStatePointers = Dict[int, Dict[AP, StatePointers]]  # {sort: {AP: APStates}}
+
+
+OSType = Dict[int, List[Set[int]]]  # {sort: [{states}]}
+TSType = Dict[int, Dict[PlanningObject, List[AP]]]  # {sort: {obj: [AP]}}
+
+
+@dataclass
+class HSIndex:
+    B: AP
```

Comment:
I assume these var names come from the paper? Would be good to have a legend in comments at the top describing what they are.